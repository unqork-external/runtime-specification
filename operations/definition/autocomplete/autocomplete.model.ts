import { DiscriminatorValue, Required } from '@tsed/schema'

import { AutocompleteOperationOptions } from './autocomplete.options'
import { Alpha } from '../../../../decorators'
import { AutogenDisabled } from '../../decorators/autogenDisabled.decorator'
import { OperationTypes } from '../../enums/operation-types.enum'
import { Operation } from '../../interface/operations.interface'

@DiscriminatorValue(OperationTypes.AUTOCOMPLETE)
@Alpha()
@AutogenDisabled()
export class AutocompleteOperation extends Operation<OperationTypes.AUTOCOMPLETE> {
  public type = OperationTypes.AUTOCOMPLETE as const

  @Required()
  public options: AutocompleteOperationOptions = new AutocompleteOperationOptions()
}
