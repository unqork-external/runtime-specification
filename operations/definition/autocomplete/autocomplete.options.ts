import { Default, Description, Required } from '@tsed/schema'

class SearchData {
  @Required()
  targetKey: string

  @Default('value')
  property?: string = 'value'
}

class Query {
  @Required()
  @Description(`
    Use JSONPath to search through the data for your intended options.
  `)
  jsonPath: string
}

class Output {
  @Required()
  targetKey: string

  @Default('value')
  property?: string = 'value'
}

export class AutocompleteOperationOptions {
  @Required()
  targetKey: string

  @Default('value')
  property?: string = 'value'

  @Required()
  @Description(`
    The data intended to be searched. The intended search data may be Array or JSON.
    Use the Query object in order to select the relevant data to look through.
  `)
  searchData: SearchData

  @Required()
  @Description(`
    The query string used to inspect the data. Only supports JSON Path right now.
  `)
  query: Query

  @Required()
  @Description(`
    \`outputData\` describes where the data should be stored after autocomplete finishes its fuzzy matches.
  `)
  outputData: Output
}
