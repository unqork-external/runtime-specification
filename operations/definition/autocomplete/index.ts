export * from './autocomplete.model'
export * from './autocomplete.options'
