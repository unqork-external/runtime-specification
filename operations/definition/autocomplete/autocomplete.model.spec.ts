import { AutocompleteOperation } from './autocomplete.model'
import { generateSchemaAndValidate } from '../../../../utilities/generateSchemaAndValidate/generateSchemaAndValidate'

describe('Operation Model: Autocomplete Tests', () => {
  let validate, schema
  beforeAll(() => {
    const response = generateSchemaAndValidate(AutocompleteOperation)
    validate = response.validate
    schema = response.schema
  })

  it('should match this JSON Schema', () => {
    expect(schema).toMatchSnapshot()
  })
})
