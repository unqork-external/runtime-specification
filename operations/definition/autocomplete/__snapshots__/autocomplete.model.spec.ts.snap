// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Operation Model: Autocomplete Tests should match this JSON Schema 1`] = `
{
  "definitions": {
    "AutocompleteOperationOptions": {
      "properties": {
        "inputs": {
          "description": "
    Inputs to be used within formulas in other operation options
  ",
          "items": {
            "$ref": "#/definitions/InputRef",
          },
          "type": "array",
        },
        "outputData": {
          "$ref": "#/definitions/Output",
        },
        "property": {
          "default": "value",
          "type": "string",
        },
        "query": {
          "$ref": "#/definitions/Query",
        },
        "searchData": {
          "$ref": "#/definitions/SearchData",
        },
        "shouldExecute": {
          "description": "
    Conditionally execute this operation based on a formula.
    Can set to false to temporarily disabled when debugging
  ",
          "type": [
            "string",
            "boolean",
          ],
        },
        "targetKey": {
          "description": "Key or path that this operation should target or operate upon",
          "examples": [
            "myComponent",
            "panel.*",
            "grid.row(#).col(name)",
          ],
          "minLength": 1,
          "type": "string",
        },
      },
      "required": [
        "targetKey",
        "searchData",
        "query",
        "outputData",
      ],
      "type": "object",
    },
    "InputRef": {
      "description": "Inputs can be used to utlize dynamic values from other state within the runtime
  within a particular execution of an operation",
      "properties": {
        "alias": {
          "description": "Optionally alias resolved value for use in formulas",
          "type": "string",
        },
        "propertyPath": {
          "description": "Path to property on target to resolve",
          "type": "string",
        },
        "required": {
          "description": "If true, operation won't execute if targeted property has no value",
          "type": "boolean",
        },
        "silent": {
          "description": "Not used in later formulas, only used to determine if operation should execute",
          "type": "boolean",
        },
        "targetKey": {
          "description": "Key of target this input depends on",
          "minLength": 1,
          "type": "string",
        },
      },
      "required": [
        "targetKey",
      ],
      "type": "object",
    },
    "Output": {
      "properties": {
        "property": {
          "default": "value",
          "type": "string",
        },
        "targetKey": {
          "minLength": 1,
          "type": "string",
        },
      },
      "required": [
        "targetKey",
      ],
      "type": "object",
    },
    "Query": {
      "properties": {
        "jsonPath": {
          "description": "
    Use JSONPath to search through the data for your intended options.
  ",
          "minLength": 1,
          "type": "string",
        },
      },
      "required": [
        "jsonPath",
      ],
      "type": "object",
    },
    "SearchData": {
      "properties": {
        "property": {
          "default": "value",
          "type": "string",
        },
        "targetKey": {
          "minLength": 1,
          "type": "string",
        },
      },
      "required": [
        "targetKey",
      ],
      "type": "object",
    },
  },
  "properties": {
    "creatorSummary": {
      "description": "A detailed summary of the operation",
      "type": "string",
    },
    "name": {
      "description": "Name of the operation",
      "type": "string",
    },
    "options": {
      "$ref": "#/definitions/AutocompleteOperationOptions",
    },
    "type": {
      "const": "AUTOCOMPLETE",
      "examples": [
        "AUTOCOMPLETE",
      ],
      "type": "string",
    },
  },
  "required": [
    "options",
  ],
  "type": "object",
}
`;
